### PR TITLE
Make improvements to income tax-related variables

### LINKED ISSUE #890: Make improvements to income tax-related variables

### PR DESCRIPTION
**Pull request by anth-volk**
Fixes #890 

This PR restructures many of the variables previously within the `income_tax` variable folder. It also restructures their corresponding tests. It adds reference material for many variables and changes the output type of `pays_scottish_income_tax` to Boolean.

This does not yet address potential issues with said variables, which will come in a later PR.

### CODE REVIEW

File: policyengine_uk/variables/gov/hmrc/pensions/pension_contributions_relief.py
Comment (PavelMakarchuk) [inline policyengine_uk/variables/gov/hmrc/pensions/pension_contributions_relief.py]: make parameter

File: policyengine_uk/variables/gov/hmrc/pensions/pension_contributions_relief.py
Comment (PavelMakarchuk) [inline policyengine_uk/variables/gov/hmrc/pensions/pension_contributions_relief.py]: split up the individual calculations

File: policyengine_uk/variables/gov/hmrc/income_tax/liability/dividend_income_tax.py
Comment (PavelMakarchuk) [inline policyengine_uk/variables/gov/hmrc/income_tax/liability/dividend_income_tax.py]: can this be negative?

File: policyengine_uk/variables/gov/hmrc/income_tax/earned_income_tax.py
Comment (PavelMakarchuk) [inline policyengine_uk/variables/gov/hmrc/income_tax/earned_income_tax.py]: can we make a bool parameter to avoid duplicating code and just create an if statement for the use of  pre or post starter parameter

File: policyengine_uk/variables/gov/hmrc/income_tax/earned_income_tax.py
Comment (PavelMakarchuk) [inline policyengine_uk/variables/gov/hmrc/income_tax/earned_income_tax.py:15]: not sure which formula supersedes which formula

File: policyengine_uk/variables/gov/hmrc/income_tax/bracketized_earned_income/earned_taxable_income.py
Comment (PavelMakarchuk) [inline policyengine_uk/variables/gov/hmrc/income_tax/bracketized_earned_income/earned_taxable_income.py]: can we make this a list parameter?

File: policyengine_uk/variables/gov/hmrc/income_tax/bracketized_earned_income/earned_taxable_income.py
Comment (PavelMakarchuk) [inline policyengine_uk/variables/gov/hmrc/income_tax/bracketized_earned_income/earned_taxable_income.py]:
```suggestion
        agi = person("adjusted_net_income", period)
```

File: policyengine_uk/variables/gov/hmrc/income_tax/bases/taxable_self_employment_income.py
Comment (PavelMakarchuk) [inline policyengine_uk/variables/gov/hmrc/income_tax/bases/taxable_self_employment_income.py]: can we make this a list parameter?

File: policyengine_uk/variables/gov/hmrc/income_tax/adjusted_net_income.py
Comment (PavelMakarchuk) [inline policyengine_uk/variables/gov/hmrc/income_tax/adjusted_net_income.py]: can we make this a list parameter?

File: policyengine_uk/tests/policy/integration/income.yaml
Comment (PavelMakarchuk) [inline policyengine_uk/tests/policy/integration/income.yaml]:
```suggestion
    allowances: 12_500
```

Comment (PavelMakarchuk) [reply policyengine_uk/tests/policy/integration/income.yaml]: In general we use "_" as "," for clarity with large numbers (each set threes)

File: policyengine_uk/tests/policy/integration/income.yaml
Comment (PavelMakarchuk) [inline policyengine_uk/tests/policy/integration/income.yaml]:
```suggestion
    taxed_income: 87500

- name: Savings slice of basic rate band
```

Comment (PavelMakarchuk) [reply policyengine_uk/tests/policy/integration/income.yaml]: minor but so much easier to review with spaces in between

File: policyengine_uk/variables/gov/hmrc/pensions/pension_contributions_relief.py
Comment (PavelMakarchuk) [inline policyengine_uk/variables/gov/hmrc/pensions/pension_contributions_relief.py]: can we just make this a formula?

**Review by PavelMakarchuk — CHANGES_REQUESTED** [review]
Looks great! - some minor suggestions here 

For me in general, I am fairly allergic to formula_xxx structures as I think it limits us and forces duplicate code - compared to a bool parameter, especially if the the change from one year to another is minor 

I am also a huge fan of lists as it (IMO) creates a little bit more structure and more direct referencing - also hope that eventually we will be able to adjust those in the web app as well (also easier to create reforms right now)

File: policyengine_uk/variables/gov/hmrc/income_tax/adjusted_net_income.py
Comment (PavelMakarchuk) [reply policyengine_uk/variables/gov/hmrc/income_tax/adjusted_net_income.py]: same in other places

**Comment by PavelMakarchuk** [comment]
> Looks great! - some minor suggestions here
> 
> For me in general, I am fairly allergic to formula_xxx structures as I think it limits us and forces duplicate code - compared to a bool parameter, especially if the the change from one year to another is minor
> 
> I am also a huge fan of lists as it (IMO) creates a little bit more structure and more direct referencing - also hope that eventually we will be able to adjust those in the web app as well (also easier to create reforms right now)

Per conversation with Max: feel free to make judgment calls which lists are more/less likely to be adjusted through policy changes and convert those to lists

File: policyengine_uk/variables/gov/hmrc/income_tax/bracketized_earned_income/earned_taxable_income.py
Comment (anth-volk) [reply policyengine_uk/variables/gov/hmrc/income_tax/bracketized_earned_income/earned_taxable_income.py]: To be clear, shouldn't this be `ani`, for `adjusted_net_income`? Just want to be 100% sure.

File: policyengine_uk/tests/policy/integration/income.yaml
Comment (anth-volk) [reply policyengine_uk/tests/policy/integration/income.yaml]: I'm using too much Claude haha

File: policyengine_uk/tests/policy/integration/income.yaml
Comment (anth-volk) [reply policyengine_uk/tests/policy/integration/income.yaml]: I will gradually add these into test files as I encounter them, but won't do so proactively, considering the volume of files present in the repo.

File: policyengine_uk/variables/gov/hmrc/income_tax/liability/dividend_income_tax.py
Comment (anth-volk) [reply policyengine_uk/variables/gov/hmrc/income_tax/liability/dividend_income_tax.py]: It cannot, and it likely shouldn't be, considering the variable's formula, but I'll add in a `max_()` function to prevent edge-case breakage.

File: policyengine_uk/variables/gov/hmrc/income_tax/earned_income_tax.py
Comment (anth-volk) [reply policyengine_uk/variables/gov/hmrc/income_tax/earned_income_tax.py]: Turns out we can actually remove

File: policyengine_uk/variables/gov/hmrc/income_tax/bracketized_earned_income/earned_taxable_income.py
Comment (PavelMakarchuk) [reply policyengine_uk/variables/gov/hmrc/income_tax/bracketized_earned_income/earned_taxable_income.py]: yes `ani` - just trying to avoid using constants

File: policyengine_uk/tests/policy/integration/income.yaml
Comment (PavelMakarchuk) [reply policyengine_uk/tests/policy/integration/income.yaml]: I would only do it in the test files that are adjusted in the PR -> e.g. I made the adjustments for the universal credit as I went through it. Not a high priority but good to clean up while we are in here

File: policyengine_uk/tests/policy/integration/income.yaml
Comment (PavelMakarchuk) [reply policyengine_uk/tests/policy/integration/income.yaml]: no such thing as too much

File: changelog_entry.yaml
Comment (PavelMakarchuk) [inline changelog_entry.yaml]:
```suggestion
    - Folder distribution and formatting for income tax-related variables.
    - pays_scottish_income_tax now returns a Boolean value.
    added:
    - References to many income tax provisions.
```

File: policyengine_uk/parameters/gov/hmrc/income_tax/adjusted_net_income_components.yaml
Comment (PavelMakarchuk) [inline policyengine_uk/parameters/gov/hmrc/income_tax/adjusted_net_income_components.yaml]:
```suggestion
  unit: list
```

Comment (PavelMakarchuk) [reply policyengine_uk/parameters/gov/hmrc/income_tax/adjusted_net_income_components.yaml]: generally we try to consolidate those parameter files into 
unit: list 

We used a bunch of versions over the years but this is what we settled on

File: policyengine_uk/parameters/gov/hmrc/income_tax/adjusted_net_income_components.yaml
Comment (PavelMakarchuk) [inline policyengine_uk/parameters/gov/hmrc/income_tax/adjusted_net_income_components.yaml]:
```suggestion
      href: "https://www.legislation.gov.uk/ukpga/2007/3/section/58"
```

Comment (PavelMakarchuk) [reply policyengine_uk/parameters/gov/hmrc/income_tax/adjusted_net_income_components.yaml]: `href` not `url` in all yaml

File: policyengine_uk/parameters/gov/hmrc/income_tax/adjusted_net_income_components.yaml
Comment (PavelMakarchuk) [reply policyengine_uk/parameters/gov/hmrc/income_tax/adjusted_net_income_components.yaml]: all yaml

File: policyengine_uk/parameters/gov/hmrc/income_tax/bases/taxable_self_employment_income_deductions.yaml
Comment (PavelMakarchuk) [inline policyengine_uk/parameters/gov/hmrc/income_tax/bases/taxable_self_employment_income_deductions.yaml]:
```suggestion
      url: https://www.legislation.gov.uk/ukpga/2007/3/section/64
```

Comment (PavelMakarchuk) [reply policyengine_uk/parameters/gov/hmrc/income_tax/bases/taxable_self_employment_income_deductions.yaml]: No "" around any of the href links

File: policyengine_uk/parameters/gov/hmrc/pensions/pension_contributions_relief_age_limit.yaml
Comment (PavelMakarchuk) [inline policyengine_uk/parameters/gov/hmrc/pensions/pension_contributions_relief_age_limit.yaml:6]:
```suggestion
  label: Pension contributions relief age limit
```

File: policyengine_uk/parameters/gov/hmrc/pensions/pension_contributions_relief_age_limit.yaml
Comment (PavelMakarchuk) [inline policyengine_uk/parameters/gov/hmrc/pensions/pension_contributions_relief_age_limit.yaml]:
```suggestion
description:  The pensions contributions relief is limited to filers at or above this age threshold. 
```

Comment (PavelMakarchuk) [reply policyengine_uk/parameters/gov/hmrc/pensions/pension_contributions_relief_age_limit.yaml]: change if not inclusive

File: policyengine_uk/variables/gov/hmrc/income_tax/bracketized_liability/tax_band.py
Comment (PavelMakarchuk) [inline policyengine_uk/variables/gov/hmrc/income_tax/bracketized_liability/tax_band.py:64]: Does not matter if this is negative?

File: policyengine_uk/variables/gov/hmrc/pensions/pension_contributions_relief.py
Comment (PavelMakarchuk) [inline policyengine_uk/variables/gov/hmrc/pensions/pension_contributions_relief.py]:
```suggestion
        income = add(
```

File: policyengine_uk/parameters/gov/hmrc/pensions/pension_contributions_relief_age_limit.yaml
Comment (PavelMakarchuk) [reply policyengine_uk/parameters/gov/hmrc/pensions/pension_contributions_relief_age_limit.yaml]: (Dont have to adjust everything here) But in general we try to make the descriptions as descriptive is possible, especially if there could be some ambiguity (e.g. inclusivity) we want to specify it here - especially for rates it is good to specify what rate is for -> this is what the user seems when clicking through the parameters in the app

**Review by PavelMakarchuk — CHANGES_REQUESTED** [review]
Minor formatting adj.

File: policyengine_uk/variables/gov/hmrc/income_tax/bracketized_liability/tax_band.py
Comment (nikhilwoodruff) [reply policyengine_uk/variables/gov/hmrc/income_tax/bracketized_liability/tax_band.py:64]: It's actually fine here, both zero and negatives both get put in the `NONE` band

File: policyengine_uk/parameters/gov/hmrc/pensions/pension_contributions_relief_age_limit.yaml
Comment (anth-volk) [reply policyengine_uk/parameters/gov/hmrc/pensions/pension_contributions_relief_age_limit.yaml:6]: I'm going to skip this, and instead lowercase, all of these labels pursuant to Nikhil's comment in #893.

Comment (anth-volk) [reply policyengine_uk/parameters/gov/hmrc/pensions/pension_contributions_relief_age_limit.yaml:6]: Or actually, I will just ignore until Nikhil can clarify

File: policyengine_uk/parameters/gov/hmrc/pensions/pension_contributions_relief_age_limit.yaml
Comment (anth-volk) [reply policyengine_uk/parameters/gov/hmrc/pensions/pension_contributions_relief_age_limit.yaml]: You make a good point. The exclusion is actually limited to those below this age limit, non-inclusive, so I will be clearer.

File: policyengine_uk/parameters/gov/hmrc/pensions/pension_contributions_relief_age_limit.yaml
Comment (PavelMakarchuk) [reply policyengine_uk/parameters/gov/hmrc/pensions/pension_contributions_relief_age_limit.yaml:6]: I think his comment was related to variable labels - we use capital letter across the US for parameter labels

File: policyengine_uk/parameters/gov/hmrc/pensions/pension_contributions_relief_age_limit.yaml
Comment (PavelMakarchuk) [inline policyengine_uk/parameters/gov/hmrc/pensions/pension_contributions_relief_age_limit.yaml]: generally we try to cite the legal sources as directly as possible to show what section of the legal code is cited 
e.g. here it would be:
```
- title: Finance Act 2004, s. 188 (3)(a)
```

Comment (PavelMakarchuk) [reply policyengine_uk/parameters/gov/hmrc/pensions/pension_contributions_relief_age_limit.yaml]: this helps us with making sure that we are as confident as possible in the source and that we can make future adjustments as needed

Comment (PavelMakarchuk) [reply policyengine_uk/parameters/gov/hmrc/pensions/pension_contributions_relief_age_limit.yaml]: all parameter files

File: policyengine_uk/variables/gov/hmrc/benefits/SMP.py
Comment (PavelMakarchuk) [inline policyengine_uk/variables/gov/hmrc/benefits/SMP.py]:
```suggestion
class statutory_maternity_pay(Variable):
```

Comment (PavelMakarchuk) [reply policyengine_uk/variables/gov/hmrc/benefits/SMP.py]: this might be easier to trace through when going through calculation trees

File: policyengine_uk/variables/gov/hmrc/benefits/SSP.py
Comment (PavelMakarchuk) [inline policyengine_uk/variables/gov/hmrc/benefits/SSP.py]:
```suggestion
class statutory_sick_pay(Variable):
```

File: policyengine_uk/variables/gov/hmrc/income_tax/bases/savings_income/ISA_interest_income.py
Comment (PavelMakarchuk) [inline policyengine_uk/variables/gov/hmrc/income_tax/bases/savings_income/ISA_interest_income.py]:
```suggestion
class individual_savings_account_interest_income(Variable):
```

File: policyengine_uk/variables/gov/hmrc/income_tax/bases/taxable_pension_income.py
Comment (PavelMakarchuk) [inline policyengine_uk/variables/gov/hmrc/income_tax/bases/taxable_pension_income.py]: all pension income is taxable?

**Review by PavelMakarchuk — CHANGES_REQUESTED** [review]

File: policyengine_uk/variables/gov/hmrc/income_tax/bases/taxable_pension_income.py
Comment (anth-volk) [reply policyengine_uk/variables/gov/hmrc/income_tax/bases/taxable_pension_income.py]: This only represents private pension income, so I'll rename accordingly

File: policyengine_uk/data/datasets/frs/calibration/calibrate.py
Comment (PavelMakarchuk) [inline policyengine_uk/data/datasets/frs/calibration/calibrate.py:340]: does it matter for the data processing that we used to calibrate for all of pension income and now we change it to private pensions?

File: docs/summary/VARIABLE_STATS.md
Comment (PavelMakarchuk) [inline docs/summary/VARIABLE_STATS.md:3143]: is there public pension income?

**Review by PavelMakarchuk — CHANGES_REQUESTED** [review]

File: policyengine_uk/data/datasets/frs/calibration/calibrate.py
Comment (nikhilwoodruff) [reply policyengine_uk/data/datasets/frs/calibration/calibrate.py:340]: Not really, we just have two names for the same thing- I'll adjust

File: docs/summary/VARIABLE_STATS.md
Comment (nikhilwoodruff) [reply docs/summary/VARIABLE_STATS.md:3143]: Technically it's `state_pension`

Comment (anth-volk) [reply docs/summary/VARIABLE_STATS.md:3143]: So just to hammer this out - my read of these variables was that `state_pension` refers to the UK public pension program, while `private_pension_income` (formerly `pension_income`) refers to all other programs. Am I right on this?

Comment (nikhilwoodruff) [reply docs/summary/VARIABLE_STATS.md:3143]: Yes!

File: policyengine_uk/variables/input/income.py
Comment (nikhilwoodruff) [inline policyengine_uk/variables/input/income.py:67]: Final thing before merging this- could we add a deprecation rather than a hard break here?

```
class pension_income(Variable):
  ... existing old code

class private_pension_income(Variable):
  ...
  def formula(...):
    # Don't have data, use the old pension income
    pension_income = person("pension_income", period)
    if pension_income.sum() != 0:
      logging.warn("`pension_income` is deprecated and will be removed- use `private_pension_income` instead")
      return pension_income
    else:
      return 0
```

**Review by nikhilwoodruff — APPROVED** [review]
Looks great! Just a final thing to avoid breaking things immediately with people who are inputting pension income into the model.

File: docs/summary/VARIABLE_STATS.md
Comment (anth-volk) [reply docs/summary/VARIABLE_STATS.md:3143]: Perfect! Thanks for the clarification.

File: policyengine_uk/variables/input/income.py
Comment (anth-volk) [reply policyengine_uk/variables/input/income.py:67]: Sure, will do, thanks for the detailed explanation.

Comment (anth-volk) [reply policyengine_uk/variables/input/income.py:67]: As an update, I think I've resolved this, but I'm not sure if the implementation I've added still allows for users to manually input the variable, as well.

**Comment by PavelMakarchuk** [comment]
LGTM if the pension name change does not affect the data @nikhilwoodruff

**Pull request merged**
